pkg/translator/prometheusremotewrite/metrics_to_prw.go

```diff
@@ -27,8 +27,11 @@ type Settings struct {
 }
 
 // FromMetrics converts pmetric.Metrics to Prometheus remote write format.
-func FromMetrics(md pmetric.Metrics, settings Settings) (map[string]*prompb.TimeSeries, error) {
-	c := newPrometheusConverter()
+// Only call this function when the result of the previous call is no longer needed
+// (i.e. has been exported out of process). The return values of this function depend on
+// internal state which is reset on every call.
+func (c *PrometheusConverter) FromMetrics(md pmetric.Metrics, settings Settings) (map[string]*prompb.TimeSeries, error) {
```

Comment:
Although I like the idea of reimagining the API, we have way too much activity happening in this component at the moment, and it is mostly coming from new members like myself, @edma2, and @jmichalek132. 

The OTel-Prometheus SIG group is relatively small, and I'd love to optimize contributions for community growth. Once we have enough hands, we can plan the refactoring you are proposing. By optimizing contributions for community growth, I mean fast interactions, try our best to keep PRs open for max 2 weeks (to avoid rebases), accepting imperfect PRs while encouraging improvements in follow-ups, etc.

What do you think?

Reply:
I never said anything about reimagining an API.  I said that if we're going to expose the `PrometheusConverter` type we need to think about what that exposed API should be and I submit that simply taking this helper method and making it the public API ain't it.  The API is already there, simply not exported.  If a `Reset()` method needs to be added, that's fine.  What we shouldn't do is export this helper method, which has the wrong signature and does more work than is needed in order to keep a separation between a performance optimization and the public API of this package, simply because it's the "efficient" thing to do from the view of how long PRs remain open.  That will mean a breaking change now and another breaking change when a sensible API is actually exposed.

An alternative that wouldn't involve considering the shape of a public API at all, and wouldn't involve any breaking changes to any public API, would be to move the `converterPool` out of the exporter and into the translation package.  The existing `FromMetrics` function is the only thing that needs it and it only needs to obtain a converter, use it, and return it to the pool.  As this PR currently stands the exporter never uses a `PrometheusConverter` instance for anything more than a single invocation of `FromMetrics`.  Doing it this way should make this PR considerably smaller and eliminates any public API changes.

What do you think?

Reply:
> An alternative that wouldn't involve considering the shape of a public API at all, and wouldn't involve any breaking changes to any public API, would be to move the converterPool out of the exporter and into the translation package. The existing FromMetrics function is the only thing that needs it and it only needs to obtain a converter, use it, and return it to the pool. As this PR currently stands the exporter never uses a PrometheusConverter instance for anything more than a single invocation of FromMetrics. Doing it this way should make this PR considerably smaller and eliminates any public API changes.

@Aneurysm9 This wouldn't work because the translator package (i.e. FromMetrics) doesn't know when to put the converter back in the pool. It can't do it at the end of FromMetrics because the metrics may not have been shipped off yet; putting it back in the pool may prematurely reset its state and send invalid attributes.

Reply:
Then we're back to what I said initially: expose the correct API for the `PrometheusConverter` type, don't simply shove this shim into it.  If we're going to export that type it should have the following interface:

```go
type PrometheusConverter interface {
  FromMetrics(pmetric.Metrics, Settings) error
  TimeSeries() []prompb.TimeSeries
  Reset()
}
```

It doesn't necessarily need to be an interface, and `TimeSeries()` returning a `prompb` type complicates the use case where an interface would actually make sense, but that's the shape of the exported methods I would expect.  They're all there.  Let's export them rather than perpetuating the stopgap that's in place and making it take multiple breaking changes to get there.